### PR TITLE
URP 2D Converters

### PR DESCRIPTION
**Pull request by unity-cchu**
### Purpose of this PR
The purpose of this PR is to add URP converter support for the upgrading of 2D materials and parametric lights for use with the 2D Renderer.

---
### Testing status
I have create and tested a project with sprite-default materials used by sprite renderers in prefabs, and loaded/unloaded scenes. I have tested upgrading materials created with the sprite/default shader. I have tried upgrading scenes and prefabs with parametric lights set. I have also tested to make sure that the parametric lights can still be upgraded through the inspector.

### CODE REVIEW

**Comment by unity-cchu** [comment]
Test scene

https://drive.google.com/file/d/1ouPZ9UQEDIp47pt_Nuxu_zujil-lNfds/view?usp=sharing

**Review by erikabar — APPROVED** [review]
2D QA will testing, I'm approving from URP QA side

**Comment by BoonKiangTan** [comment]
Changes verified on branch:

Version: 2021.2.0a20.2866
Revision: trunk 0bd21e314fca
Built: Thu, 03 Jun 2021 18:52:14 GMT
Package branch: 2d/Upgraders (1d216f4a5c)

- Verified upgrading materials created with sprite/default shader using URP converter
- Upgrading parametric lights in scenes and prefabs using URP converter

Issues found:
- Console errors when entering playmode with the converter tab open (not 2D or PR specific issue)
Unity_XE36zxHEcj

**Review by BoonKiangTan — APPROVED** [review]

**Review by kennytann — APPROVED** [review]
looks good

File: com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                            newMaterials[matIndex] = renderer.sharedMaterials[matIndex];
+                    }
+
+                    if (updateMaterials)
+                        renderer.sharedMaterials = newMaterials;
+                }
+            }
+        }
+
+        public override void OnInitialize(InitializeConverterContext context, Action calback)
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs:64]: Action callback, (my typo from long time ago that still lingers... sorry for that)

File: com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    context.AddAssetToConvert(desc);
+                }
+            }
+
+            calback.Invoke();
+        }
+
+        public override void OnRun(ref RunItemContext context)
+        {
+            string result = null;
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs:106]: can this not be string.empty() ?

File: com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            string ext = Path.GetExtension(context.item.descriptor.info);
+            if (ext == ".prefab")
+                result = URP2DConverterUtility.UpgradePrefab(context.item.descriptor.info, UpgradeGameObject);
+            else if (ext == ".unity")
+                URP2DConverterUtility.UpgradeScene(context.item.descriptor.info, UpgradeGameObject);
+            else if (ext == ".mat")
+                URP2DConverterUtility.UpgradeMaterial(context.item.descriptor.info, m_SpritesDefaultShader, m_SpriteLitDefaultShader);
+
+            if (result != null)
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs:115]: string.imptu() check instead?

File: com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                light.lightType = Light2D.LightType.Freeform;
+
+                float radius = light.shapeLightParametricRadius;
+                float angle = light.shapeLightParametricAngleOffset;
+                int sides = light.shapeLightParametricSides;
+
+                var angleOffset = Mathf.PI / 2.0f + Mathf.Deg2Rad * angle;
+                if (sides < 3)
+                {
+                    radius = 0.70710678118654752440084436210485f * radius;
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs:35]: magic number here. Put it as a constant and name it to something that says what it is.

File: com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            Light2D[] lights = go.GetComponentsInChildren<Light2D>();
+            foreach (Light2D light in lights)
+            {
+                if (light.lightType == Light2D.LightType.Parametric && !PrefabUtility.IsPartOfPrefabInstance(light))
+                    UpgradeParametricLight(light);
+            }
+        }
+
+        public override void OnInitialize(InitializeConverterContext context, Action calback)
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs:74]: same here, callback is misspelled (my mistake, sry)

File: com.unity.render-pipelines.universal/Editor/2D/Converter/URP2DConverterUtility.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/2D/Converter/URP2DConverterUtility.cs:3]: These two are not used I think

**Review by martint-unity — CHANGES_REQUESTED** [review]
A few comments. Nothing big.

File: com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        public override string name => "Parametric to Freeform Light Upgrade";
+        public override string info => "This will upgrade all parametric lights to freeform lights.";
+        public override int priority => - 1000;
+        public override Type container => typeof(UpgradeURP2DAssetsContainer);
+
+        List<string> m_AssetsToConvert = new List<string>();
+
+        string m_Light2DId;
+
+        public static void UpgradeParametricLight(Light2D light)
```
[the remark this answers is not in the document]
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs:22]: That code has been removed

File: com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs (hunk 2 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs]: Fixed

File: com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs (hunk 3 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Editor/2D/Converter/BuiltInToURP2DMaterialUpgrader.cs]: Fixed

File: com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs (hunk 4 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs]: Fixed

File: com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs (hunk 5 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Editor/2D/Converter/ParametricToFreeformLightUpgrader.cs]: Fixed

File: com.unity.render-pipelines.universal/Editor/2D/Converter/URP2DConverterUtility.cs (hunk 6 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Editor/2D/Converter/URP2DConverterUtility.cs]: Fixed

**Review by martint-unity — APPROVED** [review]

**Pull request merged**